### ISSUE TITLE
[E2E tests] Add wrappers to simulate UI flows programmatically

### ISSUE THREAD

**Issue by deepakpathania** [Automattic/woocommerce-payments#8032]
#### Problem statement

In the current E2E test, we sometimes have to reach a particular state as a setup for a test, which we achieve by simulating that flow via the UI. The actual test might not be affected by how this state is reached so doing them via UI takes extra time and could lead to flakiness in cases.

#### Acceptance criteria

- Introduce a set of wrappers to achieve common functionalities programmatically (CLI/API calls etc)
- Update usage in existing tests to use the new wrappers instead of UI flows.
- Bonus: correct the screenshot taking time in case of failures (happens after user is logged out sometimes so login screen is shown instead)

**Comment by haszari** [comment]
> Introduce a set of wrappers to achieve common functionalities programmatically (CLI/API calls etc)

@Jinksi and @ismaeldcom have some work in progress using Playwright to standardise and automate test setup. 

> programmatically (CLI/API calls etc)

Take care with this – ideally tests reproduce scenarios and flows in the same way merchants do. If we introduce (internal) APIs for fundamental flow steps then the tests could diverge from what merchants experience in subtle ways.

**Comment by ismaeldcom** [comment]
>Take care with this – ideally tests reproduce scenarios and flows in the same way merchants do. If we introduce (internal) APIs for fundamental flow steps then the tests could diverge from what merchants experience in subtle ways.

I agree, if we want to stick to the purpose of E2E testing, we should avoid mocking/bypassing any step in our flows. Having said that, I think we can wrap external functionality from WP/WC, for example we don't need to test product creation or add to cart.

**Comment by deepakpathania** [comment]
I agree that we shouldn't divert from the UI flow, this is however in scenarios where we test the flow individually already and that state might be a pre-requisite for another so that we don't re-test the same flows in multiple tests.

Consider a case where we test the payment flow already in a test, and now another test has to build on top of that state, so we could have wrappers for the client to reach that state for the subsequent test in this case, which is in line to what Ismael mentioned.

**Comment by haszari** [comment]
> Consider a case where we test the payment flow already in a test, and now another test has to build on top of that state, so we could have wrappers for the client to reach that state for the subsequent test in this case, which is in line to what Ismael mentioned.

Gotcha – so would you say the goal of this issue is optimisation, i.e. speeding up that second test?

**Comment by deepakpathania** [comment]
> so would you say the goal of this issue is optimisation, i.e. speeding up that second test?

Yes, that and by avoiding testing the same flows in multiple tests, reducing the overall time of the test suite.

**Comment by frosso** [comment]
The Playwright migration that this issue predates has shipped, and it covers all three asks here.

Programmatic setup wrappers now exist: `tests/e2e/utils/rest-api.ts` exposes WC REST API helpers (`createCustomer`, `recreateCustomer`, `createOrder`, `deleteWidgets`, etc.), landed in #10244. Shared flow helpers live in `utils/merchant.ts` and `utils/shopper.ts`, and auth state is set up once and reused via `auth.setup.ts` + `dependencies: ['setup']`. Tests already consume these.

The bonus screenshot item is moot too: Playwright captures failures natively (`screenshot: 'only-on-failure'` plus trace/video on retry), so the old manual post-logout screenshot logic the issue mentions is gone along with Puppeteer itself (#10352).

One thing worth calling out: the original thread was deliberately scoped to wrapping WP/WC setup (customers, orders, widgets) and reusing already-tested state, _not_ wrapping WooPayments payment flows - and the current architecture honors exactly that boundary. So I think we can close this. If there's still appetite for broader wrapper coverage, that's better tracked as a fresh, concretely-scoped issue than this 2024 placeholder.

Closing as completed by the Playwright migration.
